dali/operators/random/rng_base_cpu.h

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_RNG_BASE_CPU_H_
+#define DALI_OPERATORS_RANDOM_RNG_BASE_CPU_H_
+
+#include <random>
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/util/batch_rng.h"
+#include "dali/core/static_switch.h"
+
+#define DALI_NORMDIST_TYPES (uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, \
+                             int64_t, float16, float, double)
+
+namespace dali {
+
+DALI_SCHEMA(RNGAttr)
+    .DocStr(R"code()code")
+    .AddOptionalArg<std::vector<int>>("shape",
+      R"code(Shape of the data.)code", nullptr, true)
+    .AddOptionalArg<DALIDataType>("dtype",
+      R"code(Data type.)code", nullptr);
+
+template<typename Impl>
+class RNGBaseCPU : public Operator<CPUBackend> {
+ public:
+  ~RNGBaseCPU() override = default;
+
+ protected:
+  explicit RNGBaseCPU(const OpSpec &spec)
+      : Operator<CPUBackend>(spec),
+        rng_(spec.GetArgument<int64_t>("seed"), spec.GetArgument<DALIDataType>("batch_size")) {
+  }
+
+  Impl &This() noexcept { return static_cast<Impl&>(*this); }
+  const Impl &This() const noexcept { return static_cast<const Impl&>(*this); }
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc,
+                 const workspace_t<CPUBackend> &ws) override {
+    if (!spec_.TryGetArgument(dtype_, "dtype"))
+      dtype_ = This().DefaultDataType();
+
+    bool has_shape = spec_.ArgumentDefined("shape");
+    bool has_shape_like = spec_.NumRegularInput() == 1;
+    DALI_ENFORCE(!(has_shape && has_shape_like),
+      "Providing argument \"shape\" is incompatible with providing a shape-like input");
+
+    if (has_shape_like) {
+      shape_ = ws.template InputRef<CPUBackend>(0).shape();
+    } else if (has_shape) {
+      GetShapeArgument(shape_, spec_, "shape", ws);
+    } else {
+      shape_ = uniform_list_shape(spec_.GetArgument<DALIDataType>("batch_size"), {1});
```

Comment:
I think this should be merged after 1st variable batch size PR and use the new API, or we'll never merge #2408.

Reply:
sure